crates/jstz_core/src/js_fn.rs

```diff
@@ -0,0 +1,136 @@
+use std::{marker::PhantomData, ops::Deref};
+
+use boa_engine::{
+    object::builtins::JsFunction, value::TryFromJs, Context, JsResult, JsValue,
+};
+use boa_gc::{custom_trace, Finalize, Trace};
+
+use crate::value::IntoJs;
+
+pub trait IntoJsArgs<const N: usize> {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; N];
+}
+
+impl IntoJsArgs<0> for () {
+    fn into_js_args(self, _context: &mut Context<'_>) -> [JsValue; 0] {
+        []
+    }
+}
+
+impl<T0: IntoJs> IntoJsArgs<1> for (T0,) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 1] {
+        [self.0.into_js(context)]
+    }
+}
+
+impl<T0: IntoJs, T1: IntoJs> IntoJsArgs<2> for (T0, T1) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 2] {
+        [self.0.into_js(context), self.1.into_js(context)]
+    }
+}
+
+impl<T0: IntoJs, T1: IntoJs, T2: IntoJs> IntoJsArgs<3> for (T0, T1, T2) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 3] {
+        [
+            self.0.into_js(context),
+            self.1.into_js(context),
+            self.2.into_js(context),
+        ]
+    }
+}
+
+/// A `JsFn<T, N, I, O>` is a `JsFunction` tagged with some Rust types used to handle the `TryFromJs` and `IntoJs` conversions automatically:
+/// - `T` is the type of the `this` parameter;
+/// - `N` is the arity;
+/// - `I` is a tuple `(I1, ..., IN)` that contains the types of the parameters;
+/// - `O` is the type of the output.
+#[derive(Debug)]
+pub struct JsFn<T: IntoJs, const N: usize, I: IntoJsArgs<N>, O: TryFromJs> {
+    function: JsFunction,
+    _this_type: PhantomData<T>,
+    _inputs_type: PhantomData<I>,
+    _output_type: PhantomData<O>,
+}
+
+impl<T: IntoJs, const N: usize, I: IntoJsArgs<N>, O: TryFromJs> Finalize
+    for JsFn<T, N, I, O>
+{
+    fn finalize(&self) {}
+}
```

Comment:
Or maybe the impl is empty because `fn finalize(&self) {}` is the default, in which case, we could indeed derive it if we don't care about the superfluous `+ Finalize` requirements for the parameters.